src/C++/SSLSocketInitiator.cpp

```diff
@@ -318,83 +323,163 @@ void SSLSocketInitiator::doConnect( const SessionID& s, const Dictionary& d )
+
+    return SSL_HANDSHAKE_SUCCEDED;
 }
 
 void SSLSocketInitiator::onConnect( SocketConnector& connector, socket_handle s )
 {
-  SocketConnections::iterator i = m_pendingConnections.find( s );
-  if( i == m_pendingConnections.end() ) return;
+  getLog()->onEvent("Socket connected handle: " + std::to_string(s));
+   
+  time_t now;
+  ::time(&now);
+
+  SocketConnections::iterator i = m_pendingConnections.find(s);
+  if (i == m_pendingConnections.end()) return;
   SSLSocketConnection* pSocketConnection = i->second;
+  
+  m_pendingConnections.erase(i);
+  m_pendingSSLHandshakes[s] = pSocketConnection;
+  pSocketConnection->setHandshakeStartTime(now);
+    
+  handshakeSSLAndHandleConnection(connector, s); 
+}
 
-  bool sslHandshakeSuccess = handshakeSSL(pSocketConnection->sslObject());
+void SSLSocketInitiator::handshakeSSLAndHandleConnection(SocketConnector& connector, socket_handle s) {
+    SocketConnections::iterator i = m_pendingSSLHandshakes.find(s);
+    if (i == m_pendingSSLHandshakes.end()) return;
+    SSLSocketConnection* pSocketConnection = i->second;
 
-  if (sslHandshakeSuccess) {
-      m_connections[s] = pSocketConnection;
-      m_pendingConnections.erase(i);
-      setConnected(pSocketConnection->getSession()->getSessionID());
-      pSocketConnection->onTimeout();
-  }else
-  {
-      setDisconnected(pSocketConnection->getSession()->getSessionID());
-	  Session* pSession = pSocketConnection->getSession();
-      if (pSession)
-      {
-          pSession->disconnect();
-          setDisconnected(pSession->getSessionID());
-      }      
-      delete pSocketConnection;
-	  m_pendingConnections.erase(i);
-  }
+    SSLHandshakeStatus sslHandshakeStatus = handshakeSSL(pSocketConnection);
+
+    if (sslHandshakeStatus == SSL_HANDSHAKE_SUCCEDED) {
+        m_connections[s] = pSocketConnection;
+        m_pendingSSLHandshakes.erase(i);
+        setConnected(pSocketConnection->getSession()->getSessionID());
+        pSocketConnection->onTimeout();
+    }
+    else if (sslHandshakeStatus == SSL_HANDSHAKE_FAILED)
+    {
+        setDisconnected(pSocketConnection->getSession()->getSessionID());
+
+        Session* pSession = pSocketConnection->getSession();
+        if (pSession)
+        {
+            pSession->disconnect();
+            setDisconnected(pSession->getSessionID());
+        }
+
+        delete pSocketConnection;
+        m_pendingSSLHandshakes.erase(i);
+
+        getLog()->onEvent("Socket deleted due to ssl handshake error");
+    }
 }
 
-void SSLSocketInitiator::onWrite( SocketConnector& connector, socket_handle s )
+
+void SSLSocketInitiator::onWrite(SocketConnector& connector, socket_handle s)
 {
+    SocketConnections::iterator iPendingSSL = m_pendingSSLHandshakes.find(s);
+    if (iPendingSSL != m_pendingSSLHandshakes.end()){
+        SSLSocketConnection* pSocketConnection = iPendingSSL->second;
+        pSocketConnection->unsignal();
+        handshakeSSLAndHandleConnection(connector, s);
+        return;
```

Comment:
if we are in SSLhandshake state, and got a notification that the socket is writable, it will call handshakeSSL to resume the handshake.